File: lib/embeddings.ts
const VOYAGE_URL = "https://api.voyageai.com/v1/embeddings";
const MODEL = "voyage-3";
export const EMBEDDING_DIM = 1024;

const hasKey = !!process.env.VOYAGE_API_KEY;
export const isEmbeddingsMocked = !hasKey;

/**
 * Embeds a batch of text chunks via Voyage AI. Returns clearly-labeled
 * deterministic mock vectors when VOYAGE_API_KEY is unset so ingestion and
 * pgvector search keep working end-to-end without a real key.
 */
export async function embed(texts: string[]): Promise<number[][]> {
  if (!hasKey) {
    return texts.map(mockEmbedding);
  }

  const res = await fetch(VOYAGE_URL, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${process.env.VOYAGE_API_KEY}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ input: texts, model: MODEL, input_type: "document" }),
  });

  if (!res.ok) {
    throw new Error(`Voyage embeddings request failed: ${res.status} ${await res.text()}`);
  }

  const data = (await res.json()) as { data: { embedding: number[] }[] };
  return data.data.map((d) => d.embedding);
}

// Deterministic hash-based pseudo-embedding — not semantically meaningful,
// only good enough to keep pgvector queries and the ingest pipeline running.
function mockEmbedding(text: string): number[] {
  const vec = new Array(EMBEDDING_DIM).fill(0);
  for (let i = 0; i < text.length; i++) {
    vec[i % EMBEDDING_DIM] += text.charCodeAt(i);
  }
  const norm = Math.sqrt(vec.reduce((s, v) => s + v * v, 0)) || 1;
  return vec.map((v) => v / norm);
}

File: python-service/app/embeddings.py
"""Embeddings — a port of ``lib/embeddings.ts``.

Falls back to deterministic mock vectors when ``VOYAGE_API_KEY`` is unset so the
eval harness and tests run end-to-end without a key. Mock vectors carry no
semantic signal, so any recall/MRR measured against them is meaningless — the
harness prints a loud warning in that case.
"""

from __future__ import annotations

import math
import os
from typing import Sequence

VOYAGE_URL = "https://api.voyageai.com/v1/embeddings"
MODEL = "voyage-3"
EMBEDDING_DIM = 1024


def is_mocked() -> bool:
    """True when no Voyage key is configured, so ``embed`` returns mock vectors.

    Read at call time rather than import time so tests can monkeypatch the env.
    """
    return not os.environ.get("VOYAGE_API_KEY")


def embed(texts: Sequence[str]) -> list[list[float]]:
    """Embed a batch of texts, via Voyage when keyed and mock vectors otherwise."""
    if not texts:
        return []

    if is_mocked():
        return [mock_embedding(t) for t in texts]

    import httpx

    response = httpx.post(
        VOYAGE_URL,
        headers={
            "Authorization": f"Bearer {os.environ['VOYAGE_API_KEY']}",
            "Content-Type": "application/json",
        },
        json={"input": list(texts), "model": MODEL, "input_type": "document"},
        timeout=60.0,
    )
    if response.status_code != 200:
        raise RuntimeError(
            f"Voyage embeddings request failed: {response.status_code} {response.text}"
        )

    payload = response.json()
    return [item["embedding"] for item in payload["data"]]


def mock_embedding(text: str, dim: int = EMBEDDING_DIM) -> list[float]:
    """Deterministic hash-based pseudo-embedding.

    Not semantically meaningful — only good enough to keep the pipeline
    running without a key. Mirrors the TypeScript mock so both harnesses
    behave the same way when unkeyed.
    """
    vec = [0.0] * dim
    for i, ch in enumerate(text):
        vec[i % dim] += ord(ch)

    norm = math.sqrt(sum(v * v for v in vec)) or 1.0
    return [v / norm for v in vec]
